src/runtime/dsl.ts

```diff
@@ -0,0 +1,431 @@
+    // a * b -> fn.math["*"](a, b)
+    // a / b -> fn.math["/"](a, b)
+    hasChanged = true;
+    while(hasChanged) {
+      hasChanged = false;
+      code = code.replace(multiply, (str, left, op, right) => {
+        hasChanged = true;
+        left = this.transformBlockCode(left, functionArgs);
+        right = this.transformBlockCode(right, functionArgs);
+        strings.push(`${libArg}.math["${op}"](${left}, ${right})`)
+        return "____" + (strings.length - 1) + "____";
+      });
+    }
+    // a + b -> fn.math["+"](a, b)
+    // a - b -> fn.math["-"](a, b)
+    let add = new RegExp(`${infixParam}\\s*(\\+|\\-)\\s*${infixParam}`, "gi");
+    hasChanged = true;
+    while(hasChanged) {
+      hasChanged = false;
+      code = code.replace(add, (str, left, op, right) => {
+        hasChanged = true;
+        left = this.transformBlockCode(left, functionArgs);
+        right = this.transformBlockCode(right, functionArgs);
+        strings.push(`${libArg}.math["${op}"](${left}, ${right})`)
+        return "____" + (strings.length - 1) + "____";
+      });
+    }
+    // a > b -> fn.compare[">"](a, b)
+    let compare = new RegExp(`${infixParam}\\s*(>|>=|<|<=|!=|==)\\s*${infixParam}`, "gi");
+    hasChanged = true;
+    while(hasChanged) {
+      hasChanged = false;
+      code = code.replace(compare, (str, left, op, right) => {
+        hasChanged = true;
+        left = this.transformBlockCode(left, functionArgs);
+        right = this.transformBlockCode(right, functionArgs);
+        strings.push(`${libArg}.compare["${op}"](${left}, ${right})`)
+        return "____" + (strings.length - 1) + "____";
+      });
+    }
+
+    hasChanged = true;
+    while(hasChanged) {
+      hasChanged = false;
+      code = code.replace(/____([0-9]+)____/gi, function(str, index:string) {
+        let found = strings[parseInt(index)];
+        if(found) hasChanged = true;
+        return found || str;
+      })
+    }
+    return code;
+  }
+
+}
+
+//--------------------------------------------------------------------
+// Program
+//--------------------------------------------------------------------
+
+class Program {
+  blocks:DSLBlock[] = [];
+  runtimeBlocks:runtime.Block[] = [];
+  index:indexes.Index;
+  constructor(public name:string) {
+    this.index = new indexes.HashIndex();
+  }
+
+  block(name:string, func:any) {
+    let block = new DSLBlock(name, func);
+    this.blocks.push(block);
+    this.runtimeBlocks.push(block.block);
+  }
+
+  input(changes:runtime.Change[]) {
+    let trans = new runtime.Transaction(changes[0].transaction, this.runtimeBlocks, changes);
+    trans.exec(this.index);
+    console.log(trans.changes.map((change, ix) => `    <- ${change}`).join("\n"));
+  }
+}
+
+//--------------------------------------------------------------------
+// Testing
+//--------------------------------------------------------------------
+
+let foo = new Program("foo");
+
```

Comment:
I've spent a good portion of today getting testing back. I wanted to get this guy in so that we can just have real test instead of doing the stuff at the end of files like we are now.